File: sol/solution/0001-0100/0014. Longest Common Prefix/Solution.ex
defmodule Solution do
  @spec longest_common_prefix(strs :: [String.t]) :: String.t
  def longest_common_prefix([]), do: ""
  def longest_common_prefix([head | tail]) do
    Enum.reduce(tail, head, fn str, acc ->
      common_prefix(acc, str)
    end)
  end
  defp common_prefix(str1, str2) do
    str1
    |> String.graphemes()
    |> Enum.zip(String.graphemes(str2))
    |> Enum.take_while(fn {c1, c2} -> c1 == c2 end)
    |> Enum.map(fn {c, _} -> c end)
    |> Enum.join()
  end
end

File: sol/solution/0001-0100/0014. Longest Common Prefix/Solution.erl
-spec longest_common_prefix(Strs :: [unicode:unicode_binary()]) -> unicode:unicode_binary().
longest_common_prefix([]) -> <<>>;
longest_common_prefix([Prefix | Rest]) ->
    longest_common_prefix(Rest, Prefix).
longest_common_prefix([], Prefix) -> Prefix;
longest_common_prefix([H | T], Prefix) ->
    longest_common_prefix(T, common_prefix(Prefix, H)).
common_prefix(A, B) ->
    common_prefix(A, B, <<>>).
common_prefix(<<>>, _, Acc) -> Acc;
common_prefix(_, <<>>, Acc) -> Acc;
common_prefix(<<H1:8, T1/binary>>, <<H2:8, T2/binary>>, Acc) when H1 =:= H2 ->
    common_prefix(T1, T2, <<Acc/binary, H1>>);
common_prefix(_, _, Acc) -> Acc.